WordPress/gutenberg
# Webfonts: Add internal-only theme.json webfonts handler - WP 6.0 stopgap

Part of #40472.

Props to @aristath who pulled together the key elements of this PR.



## What?
Using the Webfonts API from PR #37140, this PR adds a `theme.json` webfonts handler (albeit funky design) to:

* handle only `theme.json` webfonts defined in `fontFace`.
* prevent access to the webfonts handling mechanisms by encapsulating all logic into one function that serves as a private container, i.e. `_wp_theme_json_webfonts_handler()`.

This is a stopgap approach.

## Why?

See the issue for the reasons.

## How?


It uses the Webfonts API from PR #37140 but smushes the code together into `$fn = function() {}` closures that are nested inside of a global function. This global function is the container that prevents access to the handler, thus ensuring the mechanisms are hidden and not accessible for extenders to interact with the handler.

## Testing Instructions


#### Set up instructions:

Using Twenty Twenty-Two theme, either:
* Option 1: Download Twenty Twenty-Two's new style variations https://github.com/WordPress/wordpress-develop/pull/2440
* Option 2: Modify the current version in WordPress Core by:
     * In its `functions.php` file, comment out lines 58 and 81 (the lines that are add inline styles using the font face styles).
     * Replace the `fontFamilies` in its `theme.json` file

```
			"fontFamilies": [
				{
					"fontFamily": "-apple-system,BlinkMacSystemFont,\"Segoe UI\",Roboto,Oxygen-Sans,Ubuntu,Cantarell,\"Helvetica Neue\",sans-serif",
					"name": "System Font",
					"slug": "system-font"
				},
				{
					"fontFamily": "\"Source Serif Pero\", serif",
					"name": "Source Serif Pero",
					"slug": "source-serif-pero",
					"fontFace": [
						{
							"fontFamily": "Source Serif Pero",
							"fontWeight": "200 900",
							"fontStyle": "normal",
							"fontStretch": "normal",
							"src": [ "file:./assets/fonts/SourceSerif4Variable-Roman.ttf.woff2" ]
						},
						{
							"fontFamily": "Source Serif Pero",
							"fontWeight": "200 900",
							"fontStyle": "italic",
							"fontStretch": "normal",
							"src": [ "file:./assets/fonts/SourceSerif4Variable-Italic.ttf.woff2" ]
						}
					]
				}
			],
```

#### How to Test

#### How to Test on the front-end

1. Activate TT2
2. Load the site's home page in the browser.
3. In the browser, right click to open the dev tools UI and then select "Inspect" menu item.
4. In the HTML inspector, open the `<head>` element to view all of the nodes within the head.
5. Search for the `<style>` element that has an ID of `wp-webfonts-inline-css`.
6. Open that element.

Expected Results:
The `<head>` should have one instance of the following HTML style element:

```hmtl
<style id="wp-webfonts-inline-css">
@font-face{font-family:"Source Serif Pro";font-style:normal;font-weight:200 900;font-display:fallback;src:local("Source Serif Pro"), url('/wp-content/themes/twentytwentytwo/assets/fonts/source-serif-pro/SourceSerif4Variable-Roman.ttf.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"Source Serif Pro";font-style:italic;font-weight:200 900;font-display:fallback;src:local("Source Serif Pro"), url('/wp-content/themes/twentytwentytwo/assets/fonts/source-serif-pro/SourceSerif4Variable-Italic.ttf.woff2') format('woff2');font-stretch:normal;}
</style>
```

This element should have 
* 2 `@font-face` declarations to match what was defined in the `theme.json` file.
* The `url` should no longer have the `file:./` placeholder in it.

#### How to Test in the back-end

1. Go to "Appearance" > Themes and make sure TT2 is activated.
2. Go to "Appearance" > Editor (beta) to load the site editor.
3. In the browser, right click to open the dev tools UI and then select "Inspect" menu item.
4. In the HTML inspector, open the `<head>` element to view all of the nodes within the head.
5. Search for the `<style>` element that has an ID of `wp-block-library-inline-css`. Open that element (see Expected Results for the CSS).
6. In the Network tab, select "Disable Cache" and "Fonts".
7. Refresh the page. Notice the font file(s) that are loaded.

Expected Results:
The `<head>` should have one instance of the following HTML style element:

```hmtl
<style id="wp-block-library-inline-css">
@font-face{font-family:"Source Serif Pro";font-style:normal;font-weight:200 900;font-display:fallback;src:local("Source Serif Pro"), url('/wp-content/themes/twentytwentytwo/assets/fonts/source-serif-pro/SourceSerif4Variable-Roman.ttf.woff2') format('woff2');font-stretch:normal;}@font-face{font-family:"Source Serif Pro";font-style:italic;font-weight:200 900;font-display:fallback;src:local("Source Serif Pro"), url('/wp-content/themes/twentytwentytwo/assets/fonts/source-serif-pro/SourceSerif4Variable-Italic.ttf.woff2') format('woff2');font-stretch:normal;}
</style>
```

This element should have 
* 2 `@font-face` declarations to match what was defined in the `theme.json` file.
* The `url` should no longer have the `file:./` placeholder in it.

## Design Considerations

* To enable testing, this PR comments out loading the Webfonts API files in the `lib/load.php` file. In doing so, this makes the PHPUnit tests fail as those files are missing.

* A mechanism is needed in Gutenberg to continue developing, testing, and maturing the Webfonts API. 
    * This PR currently hooks  `_wp_theme_json_webfonts_handler()` into `'wp_plugins'` action event, giving Gutenberg's plugin the opportunity to remove that callback and load its files.
    * Another way to accomplish this is to create a short-circuit within the container. In doing that, the container could be an unnamed static closure registered to the `'wp_plugins'` action event. Doing this would eliminate deprecating the function once the Webfonts API is merged into Core.
